### PR TITLE
Policy: create dynamic-pools policy for cri-rm

### PR DESCRIPTION
**Pull request by Yugar-1**
This PR creates a new policy for `cri-resource-manager` named `dynamic-pools` policy.
This policy can put the workload in different dynamic pools, and the CPUs of dynamic pools do not overlap with each other. These dynamic pools can be dynamically resized based on the requests of the containers and the CPU utilization of the workload in the dynamic pools.
Users can define and configure dynamic pools. A dynamic pool has a set of CPUs and a set of containers running on the CPUs. CPUs in dynamic pools can be configured, for example, by setting min and max frequencies on CPU cores and uncore.
The code and documentation of the `dynamic-pools` policy are included in this PR.

### CODE REVIEW

**Comment by jukkar** [comment]
The commits should be squashed together and commit message enhanced.
How this differs from balloons policy as the code seems to be copied from balloons policy?

**Comment by Yugar-1** [comment]
> The commits should be squashed together and commit message enhanced. How this differs from balloons policy as the code seems to be copied from balloons policy?

The commits have been squashed together.
The dynamic-pools policy allocates CPUs to pools differently than the balloons policy. The dynamic-pools policy dynamically allocates CPUs based on the requests of the containers and the CPU utilization of the workload in the dynamic pools, and can reallocate CPUs at regular intervals to make CPU allocation more balanced.
The dynamic-pools policy cannot be merged into the balloons policy, because they have conficts, which will modify the behavior of the balloon policy, and the characteristics of the balloons cannot be maintained.

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go [hunk 1]
```diff
@@ -0,0 +1,938 @@
...
+	for _, cID := range dp.ContainerIDs() {
+		if c, ok := p.cch.LookupContainer(cID); ok {
+			p.pinCpuMem(c, dp.Cpus, dp.Mems)
+		}
+	}
+}
+
+// updatePoolCpuset updates the cpuset of the dynamicPools.
+func (p *dynamicPools) updatePoolCpuset() error {
+	requestCpu, remainFree := p.calculateAllPoolRequests()
```
Comment (askervin) [inline pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go:190]: A situations to consider and include in e2e tests: none of the containers in a dynamic pool request any CPUs. (The containers may be BestEffort or they may have only resource limit.) It looks to me that in that case the container might get an empty cpuset, which will mean "can run on any CPU".

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go [hunk 2]
```diff
@@ -0,0 +1,938 @@
...
+	if !isNeedReallocate {
+		log.Info("The number of CPUs required by the pools is the same as the number of CPUs already in the pools, so there is no need to reallocate.")
+		for _, dp := range p.dynamicPools {
+			p.containerPinPool(dp)
+		}
+		return nil
+	}
+
+	for _, dp := range p.dynamicPools {
+		if dp.Def.Name == "reserved" {
```
Comment (askervin) [inline pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go:256]: Reallocation of CPUs can severely harm performance of a workload. If the workload has been running on CPU package0 (socket0) and is suddenly pinned to CPU cores on package1 (socket1), then suddenly memory accesses become very expensive.

In order to avoid unnecessary CPU reallocations, I would suggest:
- First release only the necessary amount of CPUs from a pool (current number of CPUs minus requestCpu[dp], if this is a positive number).
- Then allocate CPUs only to those pools that really need them.

The current balloons policy implements topology-aware CPU allocation, that is, it allocates CPU cores that are topologically as close to the currently allocated CPUs as possible. This is a bigger enhancement and can be left to be done in future PRs.

**Review by askervin — CHANGES_REQUESTED** [review]
Thank you very much for this impressive enhancement PR! This is a very nice add-on to CRI-RM features.

I added some comments to help you polishing the code. I would also ask you to include end-to-end tests (most importantly `test/e2e/policies.test-suite/dynamic-pools/n4c16/testNN-TESTNAME/code.var.sh` files) to ensure functional quality now and for the future.

File: docs/policy/dynamic-pools.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
+# Dynamic-Pools Policy
+
+## Overview
+
+The dynamic-pools policy puts the workload in the dynamic-pool that are disjoint CPU pools. The dynamic-pools can be resized, that is CPUs added and removed, which is based on the requests of the containers and the CPU utilization of the workload in the dynamic-pools.
+
```
Comment (changzhi1990) [inline docs/policy/dynamic-pools.md:6]: The dynamic-pools policy can put the workload into different dynamic-pools. Each dynamic-pool contains several CPUs and can be resized dynamically in terms of the specific algorithms.

File: docs/policy/dynamic-pools.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+On the premise that the CPUs in each dynamic-pool can meet the requests of the pods in the dynamic-pool, the CPUs are allocated based on the CPU utilization of the workload. Dynamic-pools policy try to keep CPU utilization balanced.
+
+ CPUs in dynamic-pools can be configured, for example, by setting min and max frequencies on CPU cores and uncore.
+
+## How It Works
+
+1. The user configures the dynamic-pool types from which the policy instantiates dynamic-pools. In addition to the dynamic-pools configured by the user, there is also a built-in dynamic-pool named shared pool.
+2. A dynamic-pool has a set of CPUs and a set of containers running on the CPUs.
+3. Every container is assigned to a dynamic-pool. Allows a container to use all CPUs of its pool and no other CPUs.
+4. Each logical CPU belongs to and belongs to at most one dynamic-pool. There cannot be CPUs that do not belong to any dynamic-pool.
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply docs/policy/dynamic-pools.md:16]: The `IdleCPUClass` parameter has been removed, please see the latest PR. The CPU will always belong to a dynamic-pool, so the `IdleCPUClass` configuration option is not needed. 
When only kube-system workloads are running on a node, except for the reserved CPUs, all others are in the shared dynamic-pool.

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go (hunk 2 above)
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go:256]: Thanks for your suggestions! 
This is indeed where the dynamic-pools policy needs to be improved. We will seriously consider your suggestions, and we will improve the performance of dynamic-pools policy in the future.

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go [hunk 5]
```diff
@@ -0,0 +1,938 @@
...
+}
+
+// updateRealCpuUsed returns cpu utilization of a dynamicPool.
+func (p *dynamicPools) updateRealCpuUsed(dp *DynamicPool) (int, error) {
+	if dp.Cpus.Size() == 0 {
+		log.Info("dynamic pool %s cpuset is 0", dp.Def.Name)
+		return 0, nil
+	}
+	cpuInfo, _ := getCpuUtilization(time.Duration(time.Second))
+	cpus := dp.Cpus.ToSlice()
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go]: Thanks for your suggestions! Already done.

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	cpuTimesStat2, err := getCpuTimesStat(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return calculateAllCpusUtilization(cpuTimesStat1, cpuTimesStat2)
+}
+
+func getCpuTimesStat(ctx context.Context) ([]cpuTimesStat, error) {
+	filename := pathProcStat("stat")
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go:50]: The option `HostRoot` is in package `resmgr`, cpu.go is in package `dyp`, and the structure `options` and variable `opt` to which `HostRoot` belongs are private to package `resmgr` (the first letter is lowercase). Can you please tell me how to call the `HostRoot` option in the package `dyp`?

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+	if cts2Total <= cts1Total {
+		return 100
+	}
+	return math.Min(100, math.Max(0, (cts2Busy-cts1Busy)/(cts2Total-cts1Total)*100))
+}
+
+func getBusyTime(cts cpuTimesStat) (float64, float64) {
+	total := cts.user + cts.system + cts.idle + cts.nice + cts.ioWait + cts.irq +
+		cts.softirq + cts.steal + cts.guest + cts.guestNice
+	if runtime.GOOS == "linux" {
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go:216]: `cts.guest` is already accounted in `cts.user` ,and `cts.guestNice` is already accounted in `cts.nice` on Linux. (Refer to the source code of htop: https://github.com/htop-dev/htop/blob/main/linux/LinuxProcessList.c)

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type cpuTimesStat struct {
+	cpu       string  `json:"cpu"`
+	user      float64 `json:"user"`
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go:18]: I thought about it, it really doesn't need to do the scaling, I have removed `clocksPerSecond`. Thank you very much for your suggestions!

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+			break
+		}
+		statLines = append(statLines, string(line))
+	}
+
+	var cpuLines []string
+	if len(statLines) < 2 {
+		return []string{""}, nil
+	}
+	for _, line := range statLines[1:] {
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go:120]: Thank you very much for your suggestions!
I skipped the total cpu line in the function `readCpuLines`, so it is indeed redundant to handle this line in Function `parseStatLine`. This part has been removed.

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (askervin) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go:48]: @Yugar-1 , I added PR #970 to provide a relatively clean way of accessing `--host-root` from a policy.

If this PR gets approved, you can do:
```
import (
        ...
        "github.com/intel/cri-resource-manager/pkg/sysfs"
        ...
)
...
func getCpuTimesStat(ctx context.Context) ([]cpuTimesStat, error) {
        filename := filepath.Join("/", sysfs.SysRoot(), "proc", "stat")
        ...
```

@marquiz , @jukkar , here's an example how PR #970 would solve this issue.^

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dyp
+
+import (
+	"testing"
+)
+
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp_test.go:22]: `updatePoolCpuset()` is difficult to test directly, so I split the two functions `isNeedReallocate()` and `calculatePoolCpuset()` from the function `updatePoolCpuset()`, and added unit tests for these two functions, which can be seen in the latest code. If you have any questions, please contact me again.

**Comment by Yugar-1** [comment]
> end-to-end tests



> Thank you very much for this impressive enhancement PR! This is a very nice add-on to CRI-RM features.
> 
> I added some comments to help you polishing the code. I would also ask you to include end-to-end tests (most importantly `test/e2e/policies.test-suite/dynamic-pools/n4c16/testNN-TESTNAME/code.var.sh` files) to ensure functional quality now and for the future.

We are going to doing end-to-end tests for dynamic-pools policy. We read the end-to-end tests of the balloons policy, but we didn't understand it very well. Do you have relevant documents that we can learn from? Thanks！

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+		return nil, dynamicPoolsError("received two CPU counts: %d != %d", len(cts1), len(cts2))
+	}
+	allCpusUtilization := make([]float64, len(cts1))
+	for i := 0; i < len(cts1); i++ {
+		allCpusUtilization[i] = calculateOneCpuUtilization(cts1[i], cts2[i])
+	}
+	return allCpusUtilization, nil
+}
+
+func pathProcStat(stat string) string {
```
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go]: Thanks for your suggestions! I have used  cri-resmgr `-host-root` and removed the `pathProcStat()` function.

File: pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Yugar-1) [reply pkg/cri/resource-manager/policy/builtin/dynamic-pools/cpu.go]: Thank you very much for your support! 
I have switched to use cri-resmgr `-host-root` instead of `HOST_PROC`.

**Comment by askervin** [comment]
Hi @Yugar-1,

I wrote you a documented example tests that introduce how to use the e2e test framework for testing the dynamic-pools policy. Please see the e2e commit in my [e2e_dynamic_pools_example](https://github.com/askervin/cri-resource-manager/tree/5Mg_e2e_dynamic_pools_example) branch.

I would advice you to start by installing e2e framework dependencies: docker and govm. You can find details in the CRI Resource Manager [Developer's guide / End-to-end tests](https://intel.github.io/cri-resource-manager/stable/docs/developers-guide/e2e-test.html) documentation.

Then, if you read through [test01 code](https://github.com/askervin/cri-resource-manager/blob/5Mg_e2e_dynamic_pools_example/test/e2e/policies.test-suite/dynamic-pools/n4c16/test01-basic-placement/code.var.sh) first, [test02 code](https://github.com/askervin/cri-resource-manager/blob/5Mg_e2e_dynamic_pools_example/test/e2e/policies.test-suite/dynamic-pools/n4c16/test02-rebalancing/code.var.sh) second, I'm sure you will get the idea and will be able to extend the test set further.

File: docs/policy/dynamic-pools.md (hunk 3 above)
Comment (Yugar-1) [reply docs/policy/dynamic-pools.md:6]: Thanks for your suggestions!
I have updated the dynamic-pools policy document, please see the latest PR.

**Comment by Yugar-1** [comment]
> 

Hi，

**Comment by Yugar-1** [comment]
> Hi @Yugar-1,
> 
> I wrote you a documented example tests that introduce how to use the e2e test framework for testing the dynamic-pools policy. Please see the e2e commit in my [e2e_dynamic_pools_example](https://github.com/askervin/cri-resource-manager/tree/5Mg_e2e_dynamic_pools_example) branch.
> 
> I would advice you to start by installing e2e framework dependencies: docker and govm. You can find details in the CRI Resource Manager [Developer's guide / End-to-end tests](https://intel.github.io/cri-resource-manager/stable/docs/developers-guide/e2e-test.html) documentation.
> 
> Then, if you read through [test01 code](https://github.com/askervin/cri-resource-manager/blob/5Mg_e2e_dynamic_pools_example/test/e2e/policies.test-suite/dynamic-pools/n4c16/test01-basic-placement/code.var.sh) first, [test02 code](https://github.com/askervin/cri-resource-manager/blob/5Mg_e2e_dynamic_pools_example/test/e2e/policies.test-suite/dynamic-pools/n4c16/test02-rebalancing/code.var.sh) second, I'm sure you will get the idea and will be able to extend the test set further.
> 
> Feel free to take the contents of this commit as a starting point for your e2e tests.

Hi, @askervin , thank you very much for your support!
With your help, I have pushed the code of dynamic-pools policy's e2e tests to the latest PR, thank you!

**Comment by askervin** [comment]
@Yugar-1 , how do you see, has this PR stabilized now for next review round, or are you working on some changes still?

**Comment by Yugar-1** [comment]
> @Yugar-1 , how do you see, has this PR stabilized now for next review round, or are you working on some changes still?

Hi @askervin , this PR is stable and the function has been completed, can it be merged first?
Aiming at the problem of low unit test coverage, we plan to create a new PR to improve the unit test coverage. What do you think?

**Comment by askervin** [comment]
Have you experienced flakiness in the e2e tests? I got a failure from the first run of test03-rebalancing.

```
...test/e2e$ reinstall_cri_resmgr=1 distro=debian-sid ./run_tests.sh policies.test-suite/dynamic-pools
...
# VM name        : n4c16-debian-sid-crirm-containerd
...
PASS dynamic-pools/n4c16/test01-basic-placement
PASS dynamic-pools/n4c16/test02-prometheus-metrics
PASS dynamic-pools/n4c16/test02-rebalancing
FAIL dynamic-pools/n4c16/test03-rebalancing
PASS dynamic-pools/n4c16/test04-reserved
PASS dynamic-pools/n4c16/test05-namespace
PASS dynamic-pools/n4c16/test06-update-configmap
PASS dynamic-pools/n4c16/test07-numa
```

Having a deeper look:
```
...test/e2e$ less policies.test-suite/dynamic-pools/n4c16/test03-rebalancing/output/run.sh.output
...
root@vm> nohup kubectl exec pod2 -- /bin/sh -c 'gzip </dev/zero >/dev/null' </dev/null >&/dev/null &
host> sleep 3
package0 die0 node0 core0 thread0 cpu00
                          thread1 cpu01
                    core1 thread0 cpu02 pod0c0/3247541
                          thread1 cpu03 pod2c0/3248925
                    mem   node0   2G    pod0c0/3247541
                                        pod2c0/3248925
              node1 core2 thread0 cpu04 pod1c0/3248637
                          thread1 cpu05 pod1c0/3248637
                    core3 thread0 cpu06 pod1c0/3248637
                          thread1 cpu07 pod1c0/3248637
                    mem   node1   2G    pod1c0/3248637
package1 die0 node2 core0 thread0 cpu08 pod1c0/3248637
                          thread1 cpu09 pod1c0/3248637
                    core1 thread0 cpu10 pod1c0/3248637
                          thread1 cpu11 pod1c0/3248637
                    mem   node2   2G    pod1c0/3248637
              node3 core2 thread0 cpu12 pod1c0/3248637
                          thread1 cpu13 pod1c0/3248637
                    core3 thread0 cpu14 pod1c0/3248637
                          thread1 cpu15 pod1c0/3248637
                    mem   node3   2G    pod1c0/3248637
### Verifying assertion 'len(cpus["pod0c0"]) == 1'

### The assertion holds.

### Verifying assertion 'len(cpus["pod1c0"]) == 1'

--> this assertion fails, len(["pod1c0"]) equals 12 at this point.
```

...in the cri-resmgr perspective the last rebalancing looked like this:

```
...test/e2e$ less policies.test-suite/dynamic-pools/n4c16/test03-rebalancing/output/cri-resmgr.output.txt
...
I: [resource-manager] rebalancing (reallocating) containers...
D: [     policy     ] rebalancing containers...
D: [     policy     ] dynamic pool shared{Cpus:2, Mems:0} request cpu 1
D: [     policy     ] dynamic pool pool1{Cpus:4-15, Mems:1,2,3} request cpu 1
D: [     policy     ] dynamic pool pool2{Cpus:3, Mems:0} request cpu 1
D: [     policy     ] sum remain free cpu 11
D: [     policy     ] dynamic pool shared cpuset: 2,  cpu utilization: 2
D: [     policy     ] dynamic pool: shared{Cpus:2, Mems:0}, weight: 2
D: [     policy     ] dynamic pool pool1 cpuset: 4-15,  cpu utilization: 46.02699261522791
D: [     policy     ] dynamic pool: pool1{Cpus:4-15, Mems:1,2,3}, weight: 46.02699261522791
D: [     policy     ] dynamic pool pool2 cpuset: 3,  cpu utilization: 3.8834951456310676
D: [     policy     ] dynamic pool: pool2{Cpus:3, Mems:0}, weight: 3.8834951456310676
D: [     policy     ] sum weight: 51.91048776085898
```

The same test failed when I ran with `distro=ubuntu-22.04`, too. Yet it passed the above verification, it failed very soon after it:
```
### Verifying assertion 'len(cpus["pod1c0"]) == 7'

--> failed, because len(cpus["pod1c0"]) was 6 instead of 7.
```

Would you check if these tests could be made more reliable? (Allow running a couple of rebalance rounds to let cpu load stabilize? Maybe `assert 6 <= len(...) <= 8` instead of requiring exactly 7?)

The coverage in e2e tests looks pretty good, so I'm ok with not increasing unit test coverage in this PR.

**Comment by Yugar-1** [comment]
> Have you experienced flakiness in the e2e tests? I got a failure from the first run of test03-rebalancing.
> 
> ```
> ...test/e2e$ reinstall_cri_resmgr=1 distro=debian-sid ./run_tests.sh policies.test-suite/dynamic-pools
> ...
> # VM name        : n4c16-debian-sid-crirm-containerd
> ...
> PASS dynamic-pools/n4c16/test01-basic-placement
> PASS dynamic-pools/n4c16/test02-prometheus-metrics
> PASS dynamic-pools/n4c16/test02-rebalancing
> FAIL dynamic-pools/n4c16/test03-rebalancing
> PASS dynamic-pools/n4c16/test04-reserved
> PASS dynamic-pools/n4c16/test05-namespace
> PASS dynamic-pools/n4c16/test06-update-configmap
> PASS dynamic-pools/n4c16/test07-numa
> ```
> 
> Having a deeper look:
> 
> ```
> ...test/e2e$ less policies.test-suite/dynamic-pools/n4c16/test03-rebalancing/output/run.sh.output
> ...
> root@vm> nohup kubectl exec pod2 -- /bin/sh -c 'gzip </dev/zero >/dev/null' </dev/null >&/dev/null &
> host> sleep 3
> package0 die0 node0 core0 thread0 cpu00
>                           thread1 cpu01
>                     core1 thread0 cpu02 pod0c0/3247541
>                           thread1 cpu03 pod2c0/3248925
>                     mem   node0   2G    pod0c0/3247541
>                                         pod2c0/3248925
>               node1 core2 thread0 cpu04 pod1c0/3248637
>                           thread1 cpu05 pod1c0/3248637
>                     core3 thread0 cpu06 pod1c0/3248637
>                           thread1 cpu07 pod1c0/3248637
>                     mem   node1   2G    pod1c0/3248637
> package1 die0 node2 core0 thread0 cpu08 pod1c0/3248637
>                           thread1 cpu09 pod1c0/3248637
>                     core1 thread0 cpu10 pod1c0/3248637
>                           thread1 cpu11 pod1c0/3248637
>                     mem   node2   2G    pod1c0/3248637
>               node3 core2 thread0 cpu12 pod1c0/3248637
>                           thread1 cpu13 pod1c0/3248637
>                     core3 thread0 cpu14 pod1c0/3248637
>                           thread1 cpu15 pod1c0/3248637
>                     mem   node3   2G    pod1c0/3248637
> ### Verifying assertion 'len(cpus["pod0c0"]) == 1'
> 
> ### The assertion holds.
> 
> ### Verifying assertion 'len(cpus["pod1c0"]) == 1'
> 
> --> this assertion fails, len(["pod1c0"]) equals 12 at this point.
> ```
> 
> ...in the cri-resmgr perspective the last rebalancing looked like this:
> 
> ```
> ...test/e2e$ less policies.test-suite/dynamic-pools/n4c16/test03-rebalancing/output/cri-resmgr.output.txt
> ...
> I: [resource-manager] rebalancing (reallocating) containers...
> D: [     policy     ] rebalancing containers...
> D: [     policy     ] dynamic pool shared{Cpus:2, Mems:0} request cpu 1
> D: [     policy     ] dynamic pool pool1{Cpus:4-15, Mems:1,2,3} request cpu 1
> D: [     policy     ] dynamic pool pool2{Cpus:3, Mems:0} request cpu 1
> D: [     policy     ] sum remain free cpu 11
> D: [     policy     ] dynamic pool shared cpuset: 2,  cpu utilization: 2
> D: [     policy     ] dynamic pool: shared{Cpus:2, Mems:0}, weight: 2
> D: [     policy     ] dynamic pool pool1 cpuset: 4-15,  cpu utilization: 46.02699261522791
> D: [     policy     ] dynamic pool: pool1{Cpus:4-15, Mems:1,2,3}, weight: 46.02699261522791
> D: [     policy     ] dynamic pool pool2 cpuset: 3,  cpu utilization: 3.8834951456310676
> D: [     policy     ] dynamic pool: pool2{Cpus:3, Mems:0}, weight: 3.8834951456310676
> D: [     policy     ] sum weight: 51.91048776085898
> ```
> 
> The same test failed when I ran with `distro=ubuntu-22.04`, too. Yet it passed the above verification, it failed very soon after it:
> 
> ```
> ### Verifying assertion 'len(cpus["pod1c0"]) == 7'
> 
> --> failed, because len(cpus["pod1c0"]) was 6 instead of 7.
> ```
> 
> Would you check if these tests could be made more reliable? (Allow running a couple of rebalance rounds to let cpu load stabilize? Maybe `assert 6 <= len(...) <= 8` instead of requiring exactly 7?)
> 
> The coverage in e2e tests looks pretty good, so I'm ok with not increasing unit test coverage in this PR.

Hi, @askervin , thank you very much for your suggestion! I have fixed the e2e tests of dynamic-pools policy. Please try again and contact me if you have any questions.

**Review by askervin — CHANGES_REQUESTED** [review]
Hi @Yugar-1, Thank you for all these improvements!

I think this large PR is very close to getting merged now.

I added a few mostly cosmetic comments. My wish is that anyone will be able to try out the dynamic-pools policy without errors just by copy-pasting the configuration samples.

All E2E tests passed with `distro={debian-sid,ubuntu-22.04,fedora-36}`.

**Comment by Yugar-1** [comment]
> Hi @Yugar-1, Thank you for all these improvements!
> 
> I think this large PR is very close to getting merged now.
> 
> I added a few mostly cosmetic comments. My wish is that anyone will be able to try out the dynamic-pools policy without errors just by copy-pasting the configuration samples.
> 
> All E2E tests passed with `distro={debian-sid,ubuntu-22.04,fedora-36}`.

Hi, @askervin, thank you so much for your comments, you noticed a lot of things that I had overlooked!
I've fixed all the bugs, thanks!

**Review by askervin — CHANGES_REQUESTED** [review]
Also, would you remove the ConfigMap header and related indentation from the "Example configuration that runs all pods in dynamic-pools." yaml in the documentation.

Please make sure that it runs fine with `cri-resmgr -force-config doc-test.cfg`. This will make it easy for readers to try out.

**Comment by Yugar-1** [comment]
> Also, would you remove the ConfigMap header and related indentation from the "Example configuration that runs all pods in dynamic-pools." yaml in the documentation.
> 
> Please make sure that it runs fine with `cri-resmgr -force-config doc-test.cfg`. This will make it easy for readers to try out.

I'm so sorry for my carelessness, I've fixed it, please check again.

File: docs/policy/dynamic-pools.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+Related configuration parameters:
+
+* `policy.ReservedResources.CPU` specifies the (number of) CPUs in the special `reserved` dynamic-pool. By default all containers in the `kube-system` namespace are assigned to the reserved dynamic-pool.
+* `policy.AvailableResources.CPU` specifies the CPUs that can be used by the policy, including `policy.ReservedResources.CPU`.
+* `cpu.classes` defines CPU classes and their parameters (such as `minFreq`, `maxFreq`, `uncoreMinFreq` and `uncoreMaxFreq`).
+
+### Example
+
+```yaml
+apiVersion: v1
```
Comment (askervin) [inline docs/policy/dynamic-pools.md:57]: The last change request from my side before approval: would you change this yaml, too, to the format that can be directly copy-pasted to a configuration file, and that it will then just work with:
```
cri-resmgr -force-config copypasted.cfg
```

**Review by askervin — CHANGES_REQUESTED** [review]

File: docs/policy/dynamic-pools.md (hunk 12 above)
Comment (Yugar-1) [reply docs/policy/dynamic-pools.md]: > The last change request from my side before approval: would you change this yaml, too, to the format that can be directly copy-pasted to a configuration file, and that it will then just work with:
> 
> ```
> cri-resmgr -force-config copypasted.cfg
> ```

Hi, @askervin , I have modified this yaml to be able to work through `cri-resmgr -force-config copypasted.cfg`, can you please check again.

**Review by askervin — APPROVED** [review]
Thank you, @Yugar-1!

LGTM.

**Comment by askervin** [comment]
Approved for CI.

@klihub , @marquiz , @jukkar , @fmuyassarov , would you like to have a look, too? :)

**Review by klihub — APPROVED** [review]
Let's merge this. Thanks Antti ! You've done a great job of reviewing it. And I apologize that I haven't had time to look into this.

I have one general comment which might be irrelevant for this PR. Since I understood so that this was forked off originally from an existing policy, it's good to eventually make sure that there is no unnecessary verbatim code duplication. If there is any substantial duplication of code which is not directly related to the actual policy logic but for instance metrics collection, or something other similar 'utility' functionality, then those maybe could be separated/exported out (in a subsequent PR) in the original policy into a subdirectory and then imported from there to this new policy. But as I said, it might not apply to this PR, if such code duplication does not exist...

**Comment by marquiz** [comment]
Yeah, lgtm this now and address issues have further enhancements in subsequent PRs. And thanks @askervin for all the reviews! I was really useless here 🙄
